### PR TITLE
Add ResNet-RS to keras.applications

### PR DESCRIPTION
**Pull request by spatil6**
**Description**

This PR adds ResNet-RS model architecture to keras.applications. refer #15780 

Revisiting ResNets: Improved Training and Scaling Strategies

ResNet-RS models are updated versions of ResNet models - [Arxiv Link](https://arxiv.org/abs/2103.07579)
The models were rewritten using Keras functional API, The model's weights are converted from [original repository](https://github.com/tensorflow/tpu/tree/acb331c8878ce5a4124d4d7687df5fe0fadcd43b/models/official/resnet/resnet_rs).

**Motiviation**

Adding ResNetRS models to keras.applications has been discussed on Tensorflow's github:
#15780 
It was mentioned that this family of models would be welcomed in keras.applications, hence this contribution.

**Usage / Changes**

The models could be used similarly to other models in keras.applications

```
from keras.applications import resnet_rs
model = resnet_rs.ResNetRS50()
```
**Todos** 
Right now, weights are hosting on @sebastian-sz GitHub, need to move on keras-team gcs buckets.

### CODE REVIEW

**Comment by haifeng-jin** [comment]
@LukeWood

**Comment by haifeng-jin** [comment]
@qlzh727

File: keras/applications/applications_load_weight_test.py [hunk 1]
```diff
@@ -71,6 +72,8 @@
         efficientnet_v2.EfficientNetV2S, efficientnet_v2.EfficientNetV2M,
         efficientnet_v2.EfficientNetV2L
     ]),
+    'resnet_rs':(resnet_rs,[resnet_rs.ResNetRS50,resnet_rs.ResNetRS101,resnet_rs.ResNetRS152,
+    resnet_rs.ResNetRS200,resnet_rs.ResNetRS270,resnet_rs.ResNetRS350,resnet_rs.ResNetRS420]),
```
Comment (fchollet) [inline keras/applications/applications_load_weight_test.py:76]: Fix indent

File: keras/applications/resnet_rs.py [hunk 2]
```diff
@@ -0,0 +1,891 @@
...
+    "config": {
+        "scale": 2.0,
+        "mode": "fan_out",
+        "distribution": "truncated_normal"
+    }
+}
+
+def Conv2DFixedPadding(filters, kernel_size, strides, name):
+    """Conv2D block with fixed padding.
+    """
```
Comment (fchollet) [inline keras/applications/resnet_rs.py:128]: Move `"""` to line above

File: keras/applications/resnet_rs.py [hunk 3]
```diff
@@ -0,0 +1,891 @@
...
+
+
+def BottleneckBlock(
+    filters: int,
+    strides: int,
+    use_projection: bool,
+    bn_momentum: float,
+    bn_epsilon: float,
+    activation: str,
+    se_ratio: float,
```
Comment (fchollet) [inline keras/applications/resnet_rs.py:249]: Most of these arguments should have default values

File: keras/applications/resnet_rs.py [hunk 4]
```diff
@@ -0,0 +1,891 @@
...
+            epsilon=bn_epsilon,
+            name="stem_batch_norm_4",
+        )(x)
+        x = layers.Activation(activation, name="stem_act_4")(x)
+        return x
+
+    return apply
+
+
+def SE(in_filters: int, se_ratio: float, name: str, expand_ratio: int = 1):
```
Comment (fchollet) [inline keras/applications/resnet_rs.py:201]: `name` should go last and default to `None`. You can then use `get_uid` to uniquify the default name if None https://keras.io/api/utils/backend_utils/#getuid-function

File: keras/applications/resnet_rs.py [hunk 5]
```diff
@@ -0,0 +1,891 @@
...
+CONV_KERNEL_INITIALIZER = {
+    "class_name": "VarianceScaling",
+    "config": {
+        "scale": 2.0,
+        "mode": "fan_out",
+        "distribution": "truncated_normal"
+    }
+}
+
+def Conv2DFixedPadding(filters, kernel_size, strides, name):
```
Comment (fchollet) [inline keras/applications/resnet_rs.py:126]: See previous comment about `name`

File: keras/applications/resnet_rs.py [hunk 6]
```diff
@@ -0,0 +1,891 @@
...
+            kernel_size=kernel_size,
+            strides=strides,
+            padding="same" if strides == 1 else "valid",
+            use_bias=False,
+            kernel_initializer=CONV_KERNEL_INITIALIZER,
+            name=name,
+        )(inputs)
+    return apply
+
+def STEM(bn_momentum: float, bn_epsilon: float, activation: str):
```
Comment (fchollet) [inline keras/applications/resnet_rs.py:143]: Inconsistent treatment of `name`: either you use it everywhere (preferred) or nowhere

File: keras/applications/resnet_rs.py [hunk 7]
```diff
@@ -0,0 +1,891 @@
...
+def BlockGroup(
+    filters,
+    strides,
+    se_ratio,
+    bn_epsilon,
+    bn_momentum,
+    num_repeats,
+    activation,
+    survival_probability: float,
+    name: str,
```
Comment (fchollet) [inline keras/applications/resnet_rs.py:350]: Same comment above `name`; most arguments above should have default values

File: keras/applications/resnet_rs.py [hunk 8]
```diff
@@ -0,0 +1,891 @@
...
+@keras_export('keras.applications.resnet_rs.ResNetRS50',
+              'keras.applications.ResNetRS50')
+def ResNetRS50(
+    include_top=True,
+    weights="imagenet",
+    classes=1000,
+    input_shape=None,
+    input_tensor=None,
+    pooling=None,
+    classifier_activation="softmax",
```
Comment (fchollet) [inline keras/applications/resnet_rs.py:623]: We should also include `include_preprocessing=True` in the signature for all publicly exported applications

File: keras/applications/resnet_rs.py [hunk 9]
```diff
@@ -0,0 +1,891 @@
...
+    model_name="resnet-rs",
+    pooling=None,
+    weights="imagenet",
+    input_tensor=None,
+    classes=1000,
+    classifier_activation: Union[str, Callable] = "softmax",
+    include_preprocessing=True
+):
+    """Build Resnet-RS model, given provided parameters.
+    Parameters:
```
Comment (fchollet) [inline keras/applications/resnet_rs.py:430]: The docstring formatting should be fixed (see other docstrings in the Keras codebase for examples)

File: keras/applications/resnet_rs.py [hunk 10]
```diff
@@ -0,0 +1,891 @@
...
+            img_input = input_tensor
+    bn_axis = 3 if backend.image_data_format() == "channels_last" else 1
+    x = img_input
+    if include_preprocessing:
+        num_channels = input_shape[bn_axis - 1]
+        if  num_channels == 3:
+            x = layers.Rescaling(scale=1. / 255)(x)
+            x = layers.Normalization(
+                mean=[0.485, 0.456, 0.406],
+                variance=[0.229**2, 0.224**2, 0.225**2],
```
Comment (fchollet) [inline keras/applications/resnet_rs.py:532]: Why use mean-standardization here? In general `Rescaling` is enough for preprocessing. Subtracting the mean does not help the model.

File: keras/applications/resnet_rs.py [hunk 11]
```diff
@@ -0,0 +1,891 @@
...
+                axis=bn_axis,
+            )(x)
+        else:
+            x = layers.Rescaling(scale=1. / 128.0, offset=-1)(x)
```
Comment (fchollet) [inline keras/applications/resnet_rs.py:536]: Why would the preprocessing logic be different in this case? This seems like a major gotcha

**Review by fchollet — CHANGES_REQUESTED** [review]
Thanks for the PR -- sorry about the review delay!

File: keras/applications/resnet_rs.py (hunk 2 above)
Comment (spatil6) [reply keras/applications/resnet_rs.py]: Done

File: keras/applications/resnet_rs.py (hunk 4 above)
Comment (spatil6) [reply keras/applications/resnet_rs.py]: Done

File: keras/applications/resnet_rs.py (hunk 6 above)
Comment (spatil6) [reply keras/applications/resnet_rs.py]: Added name.

File: keras/applications/resnet_rs.py (hunk 7 above)
Comment (spatil6) [reply keras/applications/resnet_rs.py]: Done

File: keras/applications/resnet_rs.py (hunk 8 above)
Comment (spatil6) [reply keras/applications/resnet_rs.py:695]: Done

File: keras/applications/resnet_rs.py (hunk 11 above)
Comment (spatil6) [reply keras/applications/resnet_rs.py]: It was mistake, forgot to remove earlier. Now taken care.

File: keras/applications/applications_test.py [hunk 12]
```diff
@@ -76,7 +77,12 @@
                         (regnet.RegNetY016, 888), (regnet.RegNetY032, 1512),
                         (regnet.RegNetY040, 1088), (regnet.RegNetY064, 1296),
                         (regnet.RegNetY080, 2016), (regnet.RegNetY120, 2240),
-                        (regnet.RegNetY160, 3024), (regnet.RegNetY320, 3712)]
+                        (regnet.RegNetY160, 3024), (regnet.RegNetY320, 3712),
+                        (resnet_rs.ResNetRS50,2048),(resnet_rs.ResNetRS101,2048),
```
Comment (LukeWood) [inline keras/applications/applications_test.py:81]: Thanks for the PR and your patience.  Sorry, this fell through the cracks on my end for a bit.

add spaces after commas in line 82-84

**Review by LukeWood — APPROVED** [review]

File: keras/applications/resnet_rs.py [hunk 13]
```diff
@@ -0,0 +1,919 @@
...
+            epsilon=bn_epsilon,
+            name=name + "_stem_batch_norm_4",
+        )(x)
+        x = layers.Activation(activation, name= name + "_stem_act_4")(x)
+        return x
+
+    return apply
+
+
+def SE(in_filters: int, se_ratio: float =0.25,  expand_ratio: int = 1,name = None):
```
Comment (fchollet) [inline keras/applications/resnet_rs.py:277]: There are style issues in the code, e.g. spacing around operators. Could you please run `black` on the file?

File: keras/applications/resnet_rs.py [hunk 14]
```diff
@@ -0,0 +1,919 @@
...
+    """Squeeze and Excitation block."""
+    bn_axis = 3 if backend.image_data_format() == "channels_last" else 1
+    if name is None:
+        name = "se" + str(backend.get_uid("se"))
```
Comment (fchollet) [inline keras/applications/resnet_rs.py:281]: Use a f-string and add a `_` before the counter

File: keras/applications/resnet_rs.py [hunk 15]
```diff
@@ -0,0 +1,919 @@
...
+        img_input = layers.Input(shape=input_shape)
+    else:
+        if not backend.is_keras_tensor(input_tensor):
+            img_input = layers.Input(tensor=input_tensor, shape=input_shape)
+        else:
+            img_input = input_tensor
+    bn_axis = 3 if backend.image_data_format() == "channels_last" else 1
+    x = img_input
+    if include_preprocessing:
+        x = layers.Rescaling(scale=1. / 255)(x)
```
Comment (fchollet) [inline keras/applications/resnet_rs.py:608]: Did you train the model from scratch to match this preprocessing scheme? Or were the weights extracted from somewhere?

Comment (sebastian-sz) [reply keras/applications/resnet_rs.py:608]: @fchollet The weights were converted by me, from the [original repository](https://github.com/tensorflow/tpu/tree/master/models/official/resnet/resnet_rs). The preprocessing layer should be:
```python
x = Normalization(
    mean=[0.485 * 255, 0.456 * 255, 0.406 * 255],
    variance=[(0.229 * 255) ** 2, (0.224 * 255) ** 2, (0.225 * 255) ** 2],
)(x)
``` 
to match the original preprocessing.

File: keras/applications/applications_test.py (hunk 12 above)
Comment (spatil6) [reply keras/applications/applications_test.py]: Done

File: keras/applications/resnet_rs.py (hunk 13 above)
Comment (spatil6) [reply keras/applications/resnet_rs.py]: Done

File: keras/applications/resnet_rs.py (hunk 14 above)
Comment (spatil6) [reply keras/applications/resnet_rs.py]: Done

File: keras/applications/resnet_rs.py (hunk 15 above)
Comment (spatil6) [reply keras/applications/resnet_rs.py]: Weights were converted by @sebastian-sz  👍 

I have also updated preprocessing layer.

Below you can find results for imagenet evaluation, 

<!DOCTYPE html>
Depth | Image size | Official Top 1 | With converted weights Top 1 | With converted weights Top 5 | Top 1 diff
-- | -- | -- | -- | -- | --
50 | 160 | 78.8 | 78.9 | 94.3 | +0.1
101 | 160 | 80.3 | 80.3 | 95.0 | 0.0
101 | 192 | 81.2 | 81.3 | 95.6 | +0.1
152 | 192 | 82.0 | 82.1 | 95.9 | +0.1
152 | 224 | 82.2 | 82.9 | 96.2 | +0.7
152 | 256 | 83.0 | 83.0 | 96.3 | 0.0
200 | 256 | 83.4 | 83.4 | 96.5 | 0.0
270 | 256 | 83.8 | 83.8 | 96.7 | 0.0
350 | 256 | 84.0 | 84.0 | 96.8 | 0.0
350 | 320 | 84.2 | 84.3 | 96.8 | +0.1
420 | 320 | 84.4 | 84.4 | 96.8 | 0.0

[Evaluation script](https://github.com/sebastian-sz/resnet-rs-keras/tree/main/imagenet_evaluation)

Comment (fchollet) [reply keras/applications/resnet_rs.py]: > Following are comparative imagenet evaluation results with official weights,

Thanks for the info! Did you generate these numbers from the latest code in this PR? Do you have a Colab I can check out?

File: keras/applications/applications_test.py [hunk 16]
```diff
@@ -76,7 +77,12 @@
                         (regnet.RegNetY016, 888), (regnet.RegNetY032, 1512),
                         (regnet.RegNetY040, 1088), (regnet.RegNetY064, 1296),
                         (regnet.RegNetY080, 2016), (regnet.RegNetY120, 2240),
-                        (regnet.RegNetY160, 3024), (regnet.RegNetY320, 3712)]
+                        (regnet.RegNetY160, 3024), (regnet.RegNetY320, 3712),
+                        (resnet_rs.ResNetRS50,2048), (resnet_rs.ResNetRS101,2048),
+                        (resnet_rs.ResNetRS152,2048), (resnet_rs.ResNetRS200,2048),
+                        (resnet_rs.ResNetRS270,2048), (resnet_rs.ResNetRS350,2048),
+                        (resnet_rs.ResNetRS420,2048)
```
Comment (fchollet) [inline keras/applications/applications_test.py:84]: In all added lines above, make sure to add a space after `,`

File: keras/applications/resnet_rs.py [hunk 17]
```diff
@@ -0,0 +1,955 @@
...
+            - `max` means that global max pooling will
+                be applied.
+        classes: optional number of classes to classify images into, only to be
+            specified if `include_top` is True, and if no `weights` argument is
+            specified.
+        classifier_activation: A `str` or callable. The activation function to
+            use on the "top" layer. Ignored unless `include_top=True`. Set
+            `classifier_activation=None` to return the logits of the "top" layer.
+        include_preprocessing: Boolean, whether to include the preprocessing layer
+            (`Rescaling`) at the bottom of the network. Defaults to `True`.
```
Comment (fchollet) [inline keras/applications/resnet_rs.py:188]: This needs to be updated

File: keras/applications/resnet_rs.py (hunk 9 above)
Comment (fchollet) [reply keras/applications/resnet_rs.py:523]: (Not yet done, please fix)

Comment (fchollet) [reply keras/applications/resnet_rs.py:523]: Another important note is that all these argument descriptions are redundant with the descriptions provided above. There should be a single source of truth so we don't have to copy/paste too much content when making modifications.

**Review by fchollet — COMMENTED** [review]
Thanks for the update!

File: keras/applications/applications_test.py (hunk 16 above)
Comment (spatil6) [reply keras/applications/applications_test.py]: Done.

File: keras/applications/resnet_rs.py (hunk 17 above)
Comment (spatil6) [reply keras/applications/resnet_rs.py:189]: Updated.

File: keras/applications/resnet_rs.py (hunk 15 above)
Comment (spatil6) [reply keras/applications/resnet_rs.py]: Those were earlier generated numbers.

Please refer below Colab link, where I have ran evaluation script on latest code from this PR.  Numbers are approximately matching with official repo.

https://colab.research.google.com/drive/1XoR18LoPhZ0Y1T3GXUdXvSmYlkJhSCS9?usp=sharing

**Comment by spatil6** [comment]
Thanks for review @fchollet . I have updated code according to review comment, also added Colab link of this PR code evaluation.

 https://colab.research.google.com/drive/1XoR18LoPhZ0Y1T3GXUdXvSmYlkJhSCS9?usp=sharing

**Review by fchollet — COMMENTED** [review]
Thanks for the update!

File: keras/applications/resnet_rs.py [hunk 18]
```diff
@@ -0,0 +1,959 @@
...
+            Note: if weights are 'imagenet' then Input image is normalized by ImageNet mean and standard deviation.
```
[the remark this answers is not in the document]
Comment (spatil6) [reply keras/applications/resnet_rs.py]: Done at both place.

File: keras/applications/resnet_rs.py (hunk 17 above)
Comment (spatil6) [reply keras/applications/resnet_rs.py:189]: Done

**Review by fchollet — APPROVED** [review]
LGTM, thank you

File: keras/applications/resnet_rs.py [hunk 19]
```diff
@@ -0,0 +1,959 @@
...
+    weights="imagenet",
+    input_tensor=None,
+    classes=1000,
+    classifier_activation: Union[str, Callable] = "softmax",
+    include_preprocessing=True,
+):
+    """Build Resnet-RS model, given provided parameters.
+
+    Args:
+        depth: Depth of ResNet network.
```
Comment (LukeWood) [inline keras/applications/resnet_rs.py:526]: re-order to match the order in the function parameters

Comment (spatil6) [reply keras/applications/resnet_rs.py:526]: Done.

Comment (spatil6) [reply keras/applications/resnet_rs.py:526]: @LukeWood its done. Please let me know if any additional changes required

File: keras/applications/resnet_rs.py [hunk 20]
```diff
@@ -0,0 +1,959 @@
...
+            channels, and width and height should be no smaller than 32.
+            E.g. (200, 200, 3) would be one valid value.
+        bn_momentum: Momentum parameter for Batch Normalization layers.
+        bn_epsilon: Epsilon parameter for Batch Normalization layers.
+        activation: activation function.
+        se_ratio: Squeeze and Excitation layer ratio.
+        dropout_rate: dropout rate before final classifier layer.
+        drop_connect_rate: dropout rate at skip connections.
+        include_top: whether to include the fully-connected layer at the top of
+        the network.
```
Comment (haifeng-jin) [inline keras/applications/resnet_rs.py:537]: A indent is missing on this line.

**Review by haifeng-jin — CHANGES_REQUESTED** [review]

File: keras/applications/resnet_rs.py (hunk 20 above)
Comment (spatil6) [reply keras/applications/resnet_rs.py]: Done

File: keras/applications/resnet_rs.py [hunk 21]
```diff
@@ -0,0 +1,957 @@
...
+    pooling=None,
+    weights="imagenet",
+    input_tensor=None,
+    classes=1000,
+    classifier_activation: Union[str, Callable] = "softmax",
+    include_preprocessing=True,
+):
+    """Build Resnet-RS model, given provided parameters.
+
+    Args:
```
Comment (LukeWood) [inline keras/applications/resnet_rs.py:524]: unindent docstring to match code indentation

File: keras/applications/resnet_rs.py [hunk 22]
```diff
@@ -0,0 +1,957 @@
...
+            use as image input for the model.
+        classes: optional number of classes to classify images into, only to be
+            specified if `include_top` is True, and if no `weights` argument is
+            specified.
+        classifier_activation: A `str` or callable. The activation function to
+            use on the "top" layer. Ignored unless `include_top=True`. Set
+            `classifier_activation=None` to return the logits of the "top" layer.
+        include_preprocessing: Boolean, whether to include the preprocessing layer
+            (`Rescaling`) at the bottom of the network. Defaults to `True`.
+            Note: Input image is normalized by ImageNet mean and standard deviation.
```
Comment (LukeWood) [inline keras/applications/resnet_rs.py:565]: lets change "Note:" to "Note-".  "Note:" gets picked up as a parameter name by our docstring generator

Comment (spatil6) [reply keras/applications/resnet_rs.py]: Done

File: keras/applications/resnet_rs.py (hunk 21 above)
Comment (spatil6) [reply keras/applications/resnet_rs.py]: Changed indentation from 4 spaces to 2 spaces, it should be good now. Let me know if anything, need to be done.

**Comment by LukeWood** [comment]
I've almost got this ready to merge.  Thanks for your patience!

**Comment by fchollet** [comment]
This has been merged in https://github.com/keras-team/keras/commit/a9b4cb6bba4608d44f41f2719ec50356710cadb5 and https://github.com/tensorflow/tensorflow/commit/e9c5ec2787e3dbba0cc13aed554025e4c1c4edf1 -- thanks a lot for the great contribution! 👍

**Comment by spatil6** [comment]
> This has been merged in [a9b4cb6](https://github.com/keras-team/keras/commit/a9b4cb6bba4608d44f41f2719ec50356710cadb5) and [tensorflow/tensorflow@e9c5ec2](https://github.com/tensorflow/tensorflow/commit/e9c5ec2787e3dbba0cc13aed554025e4c1c4edf1) -- thanks a lot for the great contribution! +1

@fchollet should i open new PR to update keras-io docs for information about same?

**Comment by sebastian-sz** [comment]
@shkarupa-alex 
Disclaimer: I am the author of the [Keras port repository](https://github.com/sebastian-sz/resnet-rs-keras).

Good find, I agree there is no mention of momentum in the paper and the value has been simply copied from the config file you linked. I missed that they used moving average decay. 

It could be beneficial to replace it with something more standard - e.g. maybe just leave the default batch norm momentum value?

I think you could open a separate issue regarding this problem. I don't think a closed PR will attract the needed attention.

**Comment by LukeWood** [comment]
> @shkarupa-alex Disclaimer: I am the author of the [Keras port repository](https://github.com/sebastian-sz/resnet-rs-keras).
> 
> Good find, I agree there is no mention of momentum in the paper and the value has been simply copied from the config file you linked. I missed that they used moving average decay.
> 
> It could be beneficial to replace it with something more standard - e.g. maybe just leave the default batch norm momentum value?
> 
> I think you could open a separate issue regarding this problem. I don't think a closed PR will attract the needed attention.

agreed, lets file an issue.  maybe we can fix this with the `KerasCV.models` refactor?

**Comment by sebastian-sz** [comment]
@LukeWood yes this is a minor change in code.

I think Resnet-RS models could be added to keras-cv with this fix. Do we also would like to modify `keras.applications` implementation?

**Comment by LukeWood** [comment]
If it is a minor change I see no harm in also fixing keras.applicatins - as long as it doesn't change the API (but I don't think it does)

**Comment by shkarupa-alex** [comment]
Here is a PR https://github.com/keras-team/keras/pull/16726

**Comment by Pandaaaa906** [comment]
will ResNet-RS merge into Keras V3?

**Pull request closed without merging**